### PR TITLE
Add bytecode hash in interpreter #1888

### PR DESCRIPTION
**Pull request by FredCoen**
Resolves https://github.com/bluealloy/revm/issues/1888

### CODE REVIEW

**Comment by FredCoen** [comment]
@rakita i am not sure if option is the best way to do this? Also let me know if we should set the value in the getter directly !

**Comment by rakita** [comment]
Hey, check out the ExtBytecode, it is better place for the hash. And allow ExtBytecode to be initialized with the hash, if this is regenerated every time it is a performance hit.

**Comment by FredCoen** [comment]
thanks! 

"allow ExtBytecode to be initialized with the hash" -> i introduced new_with_hash to allow this, otherwise calling the hash() method returns the hash and cashes it on the ExtByteCode. Is this what you had in mind?

For some reason it caused a regression on the ecrecover precompile perf. I am not sure why though. Do you have any idea @rakita ?

**Comment by rakita** [comment]
thanks!

> "allow ExtBytecode to be initialized with the hash" -> i introduced new_with_hash to allow this, otherwise calling the hash() method returns the hash and cashes it on the ExtByteCode. Is this what you had in mind?

Yes! It would be good for `fn hash()` to return Option as there is already a function to generate hash in `Bytecode`.

> For some reason it caused a regression on the ecrecover precompile perf. I am not sure why though. Do you have any idea @rakita ?

It is not related, it can be ignored. The test is kinda flaky.

File: crates/interpreter/src/interpreter/ext_bytecode.rs [hunk 1]
```diff
@@ -29,6 +30,26 @@ impl ExtBytecode {
         Self {
             base,
             instruction_pointer,
+            bytecode_hash: None,
+        }
+    }
+
+    pub fn new_with_hash(base: Bytecode, hash: B256) -> Self {
```
Comment (rakita) [inline crates/interpreter/src/interpreter/ext_bytecode.rs:37]:
```suggestion
    /// Creates new `ExtBytecode` with the given hash.
    pub fn new_with_hash(base: Bytecode, hash: B256) -> Self {
```

File: crates/interpreter/src/interpreter/ext_bytecode.rs [hunk 2]
```diff
@@ -29,6 +30,27 @@ impl ExtBytecode {
...
+        }
+    }
+
+    pub fn hash(&mut self) -> B256 {
+        if let Some(hash) = self.bytecode_hash {
+            hash
+        } else {
+            let hash = self.base.hash_slow();
+            self.bytecode_hash = Some(hash);
+            hash
```
Comment (rakita) [inline crates/interpreter/src/interpreter/ext_bytecode.rs:53]:
```suggestion
    pub fn hash(&mut self) -> Option<B256> {
       self.bytecode_hash
```

File: crates/interpreter/src/interpreter/ext_bytecode.rs [hunk 3]
```diff
@@ -29,6 +30,22 @@ impl ExtBytecode {
...
+        Self {
+            base,
+            instruction_pointer,
+            bytecode_hash: Some(hash),
+        }
+    }
+
+    pub fn hash(&mut self) -> Option<B256> {
+       self.bytecode_hash
         }
```
Comment (rakita) [inline crates/interpreter/src/interpreter/ext_bytecode.rs:49]:
```suggestion
```

**Comment by FredCoen** [comment]
hey i removed the cached hash test. Should i add anything else? CI passes where do you see it being flaky on your side locally?

**Comment by rakita** [comment]
> hey i removed the cached hash test. Should i add anything else? CI passes where do you see it being flaky on your side locally?

Perfect!

The recovery test is flaky as nothing is changed with precompiles but it shows a diff, we can ignore it.

Have added the ExtBytecode::new_with_hash usege and will PR merge after CI

**Review by rakita — APPROVED** [review]
lgtm!

**Pull request merged**